### PR TITLE
New block types

### PR DESCRIPTION
**Pull request by MarcoTUM**

### CODE REVIEW

File: Assets/Scripts/PlayScene/PlayScene.cs [hunk 1]
```diff
@@ -54,8 +54,58 @@ private IEnumerator PlayerDeath()
...
+    {
+        StartCoroutine(ForceOutbreak(chargeTime, outbreakRadius, forceOutbreak));
+    }
+
+    private IEnumerator ForceOutbreak(float chargeTime, float outbreakRadius, ParticleSystem forceOutbreak)
+    {
+        yield return new WaitForSeconds(chargeTime);
+
+        // Get the player script (assuming there is always only one player in the scene)
+        Player player = GameObject.FindObjectOfType<Player>();
```
Comment (lee17017) [inline Assets/Scripts/PlayScene/PlayScene.cs]: Use Gamemaster.Instance().GetPlayer()

File: Assets/Scripts/PlayScene/PlayScene.cs [hunk 2]
```diff
@@ -54,8 +54,58 @@ private IEnumerator PlayerDeath()
...
+        Collider[] hitColliders = Physics.OverlapSphere(player.transform.position, outbreakRadius);
+
+        // Instantiate force outbreak particle effect
+        Instantiate(forceOutbreak, player.transform.position, Quaternion.identity);
+
+        // For each collider destroy the block
+        foreach (Collider hitCollider in hitColliders)
+        {
+            if (hitCollider != null && !hitCollider.gameObject.CompareTag("Player"))
+            {
```
Comment (lee17017) [inline Assets/Scripts/PlayScene/PlayScene.cs:98]: Why not just if hitCollider.HasComponent<Block>()?

File: Assets/Scripts/Blocks/RestoreableBlock.cs [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+
+    [SerializeField]
+    private Material restoredMaterial;
+
+    private Material restoreableMaterial;
+
+    #region Initialization / Destruction
+    public override void InitializeBlock(Block_Data data)
+    {
+        gameObject.GetComponent<Collider>().enabled = false;
```
Comment (lee17017) [inline Assets/Scripts/Blocks/RestoreableBlock.cs]: Use gameObject.GetComponent<Collider>().enabled = SceneManager.GetActiveScene().name != SceneDictionary.Play;
and remove the deleteBlockButton

File: Assets/Scripts/Blocks/BlockData/RespawnBlock_Data.cs [hunk 4]
```diff
@@ -0,0 +1,21 @@
+﻿using System.Collections;
+using System.Collections.Generic;
+using UnityEngine;
+
+[System.Serializable]
+public class RespawnBlock_Data : Block_Data
+{
+    public override BlockType BlockType => BlockType.Respawn;
+
+    private float respawnTime;
```
Comment (lee17017) [inline Assets/Scripts/Blocks/BlockData/RespawnBlock_Data.cs:10]: Is it really necessary to save the spawnTime? it should be the the same through all the blocks, right?

File: Assets/Scripts/PlayScene/PlayScene.cs [hunk 5]
```diff
@@ -54,8 +54,58 @@ private IEnumerator PlayerDeath()
         player.gameObject.SetActive(true);
         running = true;
     }
-    
 
-  
+    #region BlockHelperFunctions
 
+    public void RespawnRespawnBlocks(RespawnBlock respawnBlock, float respawnTime)
```
Comment (lee17017) [inline Assets/Scripts/PlayScene/PlayScene.cs:59]: Should in my opinion be handled within the respawnBlock

**Review by lee17017 — CHANGES_REQUESTED** [review]
Color of block lock block reverts when they are unlocked.

RespawnBlocks need to check if player is currently on their position(otherwise respawn on player)

Charge blocks not working properly -> Still can destroy blocks and energy release is kinda wimpy(cant see if it actually does anything)

RestoreBlocks also need to test if player is on them(or we forbid to place blocks close to button)

File: Assets/Scripts/PlayScene/PlayScene.cs (hunk 1 above)
Comment (MarcoTUM) [reply Assets/Scripts/PlayScene/PlayScene.cs]: Ok

File: Assets/Scripts/PlayScene/PlayScene.cs (hunk 2 above)
Comment (MarcoTUM) [reply Assets/Scripts/PlayScene/PlayScene.cs:98]: There is no such function...

File: Assets/Scripts/Blocks/BlockData/RespawnBlock_Data.cs (hunk 4 above)
Comment (MarcoTUM) [reply Assets/Scripts/Blocks/BlockData/RespawnBlock_Data.cs]: Ok

File: Assets/Scripts/Blocks/RestoreableBlock.cs (hunk 3 above)
Comment (MarcoTUM) [reply Assets/Scripts/Blocks/RestoreableBlock.cs]: Ok

File: Assets/Scripts/PlayScene/PlayScene.cs (hunk 5 above)
Comment (MarcoTUM) [reply Assets/Scripts/PlayScene/PlayScene.cs:59]: The coroutine stops if you put the code within the respawnBlock and I don't think just disabling the collider is a better idea...

File: Assets/Scripts/Blocks/BlockData/RespawnBlock_Data.cs (hunk 4 above)
Comment (MarcoTUM) [reply Assets/Scripts/Blocks/BlockData/RespawnBlock_Data.cs:10]: No only the respawnBlock has a respawnTime, this time is the time that passes until the respawnBlock reapears.

**Review by lee17017 — APPROVED** [review]

**Pull request merged**